### PR TITLE
New language: Spice

### PR DESCRIPTION
**Pull request by marcauberer**
Add support for the Spice programming language.

Spice (`spicelang`) is a simple, AOT compiled, LLVM-based systems language with focus on practicality and performance.

**Key features of Spice**
- simplicity - see what you get aproach with batteries included
- performance: As Spice uses LLVM as its primary backbone, it competes with the performance of other LLVM-based compilers
- enhanced safety: Spice implements an enhanced safety mechanism, which includes builtin heap type with auto-free and more
- cross-compile support: Currently x86_64, AArch64 and wasm32 are supported cross-compilation targets, but more to come
- Pretty good C and C++ interop

**Links**
- Source code: https://github.com/spicelang/spice
- Website: https://spicelang.com
- Playground (own CE instance): https://play.spicelang.com
- Infra PR: https://github.com/compiler-explorer/infra/pull/1229

**Things this PR addresses**
- General support for Spice
- Three code examples
- Spice mode (Syntax highlighting)
- Opt pipeline support
- LLVM mca and osaca support

### CODE REVIEW

**Comment by marcauberer** [comment]
The pipeline error seems to be a hiccup (timeout). Tests run fine locally.

File: etc/config/spice.amazon.properties [hunk 1]
```diff
@@ -0,0 +1,40 @@
+compilers=&spice
+defaultCompiler=spice01902
+
+group.spice.compilers=spice01902
+group.spice.demangler=/opt/compiler-explorer/gcc-12.1.0/bin/c++filt
+group.spice.objdumper=/opt/compiler-explorer/gcc-12.1.0/bin/objdump
```
Comment (dkm) [inline etc/config/spice.amazon.properties:6]: you may want to use 13.2 here. Our versioning could be better, but.... :)

File: etc/config/spice.amazon.properties [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+group.spice.baseName=spice
+group.spice.compilerType=spice
+group.spice.versionFlag=--version
+group.spice.licenseLink=https://github.com/spicelang/spice/blob/main/LICENSE
+group.spice.licenseName=The MIT License
+group.spice.licensePreamble=Copyright (c) 2021-2024 Marc Auberer
+
+compiler.spice01902.exe=/opt/compiler-explorer/spice-0.19.2/spice
+compiler.spice01902.semver=0.19.2
+compiler.spice01902.name=spice 0.19.2
```
Comment (dkm) [inline etc/config/spice.amazon.properties:17]: you don't need the .name field. It should be correctly compited using semVer + baseName.

File: lib/compilers/spice.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2023, Marc Auberer
```
Comment (dkm) [inline lib/compilers/spice.ts:1]: we're in 2024 now, so maybe you can bump... But that's not critical

**Review by dkm — CHANGES_REQUESTED** [review]
minor remarks while doing a drive by review

File: examples/spice/default.spice [hunk 4]
```diff
@@ -0,0 +1,8 @@
+f<int> fibo(int n) {
```
Comment (junlarsen) [inline examples/spice/default.spice:1]: Maybe it's best to keep the example consistent to a square function like the other examples?

Comment (marcauberer) [reply examples/spice/default.spice]: Done.

**Comment by marcauberer** [comment]
Any updates on this? Any more objections?

**Comment by partouf** [comment]
lgtm, will merge this weekend

**Comment by dkm** [comment]
@marcauberer can you rebase this PR, to see if your labeler fix works?

**Comment by dkm** [comment]
```
 The configuration file (path: .github/labeler.yml) was not found locally, fetching via the api
Error: Error: The "changed-files" section must have a valid config structure. Please read the action documentation for more information
Error: The "changed-files" section must have a valid config structure. Please read the action documentation for more information
```
Erm. :D

**Comment by dkm** [comment]
The Labeler does not fail, but it doesn't seem to add the spice label here... But maybe because it's not merged, not sure...

**Review by dkm — APPROVED** [review]

**Comment by marcauberer** [comment]
@partouf Ping - What's the status?

**Comment by dkm** [comment]
> @partouf Ping - What's the status?

The status is that we still need to merge it, test, and deploy. It can take a few days, depending on the time we have :)
We'll let you know when it is merged, tested and finally live :)

**Comment by mattgodbolt** [comment]
I'm happy to merge and deal with any remaining problems as they come up! Thanks all!

**Comment by mattgodbolt** [comment]
Merged and in testing this happens:
```
terminate called after throwing an instance of 'spice::compiler::LinkerError'
  what():  [Error|Linker] Linker not found: Please check if you have installed clang and added it to the PATH variable
Program terminated with signal: SIGSEGV
Compiler returned: 139
```
I'm going to go ahead and push to production and we can debug there though

**Comment by marcauberer** [comment]
Yes, Spice uses the lld linker as default linker and clang to invoke it? Seems like no clang version can be found in the PATH env var.
Can we add the latest clang binary to the path somehow or should I patch the compiler to search it via another env var?

**Comment by mattgodbolt** [comment]
I wonder why it was trying to link: the first thing I tried was just an assembly-only (but maybe that's how it works?)

> Can we add the latest clang binary to the path somehow

Yes! I had to search about a bit but e.g. for `hylo` we set:
```
# hylo needs to be able to find clang++ for linking
group.hylo.extraPath=/opt/compiler-explorer/clang-15.0.0/bin
```

So it looks like you could add similar config and point at whichever version of clang makes the most sense!

**Comment by marcauberer** [comment]
>I wonder why it was trying to link: the first thing I tried was just an assembly-only (but maybe that's how it works?)

@mattgodbolt This was intentional before, but not really ideal for CE. Should be fixed with #6170

**Pull request merged**